File: app/src/main/java/is/digital/interviewskeleton/QuickUtils.java
package is.digital.interviewskeleton;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONObject;

import java.util.Iterator;

import okhttp3.FormBody;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

class QuickUtils {
    public static final String HOST = "https://api-nodejs-todolist.herokuapp.com";

    //--------- SETTINGS ----------------------
    private static SharedPreferences sharedPrefs;
    private static SharedPreferences.Editor sharedPrefsEditor;

    public static void init(Context context) {
        sharedPrefs = context.getSharedPreferences("SHARED_PREFS", Context.MODE_PRIVATE);
        sharedPrefsEditor = sharedPrefs.edit();
    }

    public interface QuickCallbackInterface {
        void onSuccess(String response);

        void onError(Exception e);
    }

    public static void getResponse(final Activity activity, final String url, final QuickCallbackInterface callback) {
        if (BuildConfig.DEBUG)
            Log.d("DEBUG", "getResponse: " + url);
        new Thread() {
            public void run() {
                try {
                    OkHttpClient client = new OkHttpClient.Builder().build();
                    Request request = new Request.Builder()
                            .addHeader("Authorization", "Bearer " + BuildConfig.BEARER_TOKEN)
                            .url(HttpUrl.parse(url))
                            .build();
                    Response response = client.newCall(request).execute();

                    final String responseString = response.body().string();

                    if (BuildConfig.DEBUG)
                        Log.d("DEBUG", "response: " + responseString);
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            callback.onSuccess(responseString);
                        }
                    });
                } catch (final Exception e) {
                    e.printStackTrace();
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            callback.onError(e);
                        }
                    });
                }
            }
        }.start();
    }

    public static void getPostResponse(final Activity activity, final String url, final JSONObject params, final QuickCallbackInterface callback) {
        if (BuildConfig.DEBUG)
            Log.d("DEBUG", "getPostResponse: " + url + " params: " + params);
        new Thread() {
            public void run() {
                try {
                    OkHttpClient client = new OkHttpClient.Builder()
                            .build();
                    FormBody.Builder bodyBuilder = new FormBody.Builder();
                    Iterator<String> iter = params.keys();
                    while (iter.hasNext()) {
                        String key = iter.next();
                        bodyBuilder.add(key, params.getString(key));
                    }
                    RequestBody body = bodyBuilder.build();
                    Request request = new Request.Builder()
                            .addHeader("Accept", "application/json")
                            .addHeader("Authorization", "Bearer " + BuildConfig.BEARER_TOKEN)
                            .url(url)
                            .post(body)
                            .build();
                    Response response = client.newCall(request).execute();
                    final String responseString = response.body().string();
                    if (BuildConfig.DEBUG)
                        Log.d("DEBUG", "response: " + responseString);

                    if(activity==null)
                        callback.onSuccess(responseString);
                    else {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                callback.onSuccess(responseString);
                            }
                        });
                    }
                } catch (final Exception e) {
                    e.printStackTrace();
                    if(activity==null)
                        callback.onError(e);
                    else {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                callback.onError(e);
                            }
                        });
                    }
                }
            }
        }.start();
    }
}

File: app/src/main/java/is/digital/interviewskeleton/QuickUtils.kt
package `is`.digital.interviewskeleton

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import okhttp3.FormBody
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import org.json.JSONObject

internal object QuickUtils {
    const val HOST = "https://api-nodejs-todolist.herokuapp.com"

    //--------- SETTINGS ----------------------
    private lateinit var sharedPrefs: SharedPreferences
    private lateinit var sharedPrefsEditor: SharedPreferences.Editor
    fun init(context: Context) {
        sharedPrefs = context.getSharedPreferences("SHARED_PREFS", Context.MODE_PRIVATE)
        sharedPrefsEditor = sharedPrefs.edit()
    }

    fun getResponse(activity: Activity, url: String, callback: QuickCallbackInterface) {
        if (BuildConfig.DEBUG) Log.d("DEBUG", "getResponse: $url")
        object : Thread() {
            override fun run() {
                try {
                    val client = OkHttpClient.Builder().build()
                    val request = Request.Builder()
                        .addHeader("Authorization", "Bearer " + BuildConfig.BEARER_TOKEN)
                        .url(url.toHttpUrlOrNull()!!)
                        .build()
                    val response = client.newCall(request).execute()
                    val responseString = response.body!!.string()
                    if (BuildConfig.DEBUG) Log.d("DEBUG", "response: $responseString")
                    activity.runOnUiThread { callback.onSuccess(responseString) }
                } catch (e: Exception) {
                    e.printStackTrace()
                    activity.runOnUiThread { callback.onError(e) }
                }
            }
        }.start()
    }

    fun getPostResponse(activity: Activity?, url: String, params: JSONObject, callback: QuickCallbackInterface) {
        if (BuildConfig.DEBUG) Log.d("DEBUG", "getPostResponse: $url params: $params")
        object : Thread() {
            override fun run() {
                try {
                    val client = OkHttpClient.Builder()
                        .build()
                    val bodyBuilder = FormBody.Builder()
                    val iter = params.keys()
                    while (iter.hasNext()) {
                        val key = iter.next()
                        bodyBuilder.add(key, params.getString(key))
                    }
                    val body: RequestBody = bodyBuilder.build()
                    val request = Request.Builder()
                        .addHeader("Accept", "application/json")
                        .addHeader("Authorization", "Bearer " + BuildConfig.BEARER_TOKEN)
                        .url(url)
                        .post(body)
                        .build()
                    val response = client.newCall(request).execute()
                    val responseString = response.body!!.string()
                    if (BuildConfig.DEBUG) Log.d("DEBUG", "response: $responseString")
                    if (activity == null) callback.onSuccess(responseString) else {
                        activity.runOnUiThread(Runnable { callback.onSuccess(responseString) })
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    if (activity == null) callback.onError(e) else {
                        activity.runOnUiThread(Runnable { callback.onError(e) })
                    }
                }
            }
        }.start()
    }

    interface QuickCallbackInterface {
        fun onSuccess(response: String?)
        fun onError(e: Exception)
    }
}
